File: src/core/bus.ts
import { category, object, outcome, routeOf, type Board, type BotEvent, type Doc, type Outcome } from './types.js';
import { QUALITY_KINDS } from './events.js';

export interface Listener {
  id: string; business: string; description: string; routes: readonly string[];
  handler: (event: BotEvent) => Doc | void | Promise<Doc | void>;
  exclude?: readonly string[]; providers?: readonly string[]; repositories?: readonly string[];
  enabled?: boolean; mode?: 'active' | 'noop';
}
/** Shell-style route selectors, including ?, [abc], and [!abc]. No executable regex input. */
export function glob(value: string, pattern: string): boolean {
  let regex = '^';
  for (let i = 0; i < pattern.length; i++) {
    const c = pattern[i];
    if (c === '*') regex += '.*';
    else if (c === '?') regex += '.';
    else if (c === '[' && pattern.indexOf(']', i + 1) > i + 1) {
      const end = pattern.indexOf(']', i + 1);
      let content = pattern.slice(i + 1, end).replaceAll('\\', '\\\\');
      if (content.startsWith('!')) content = '^' + content.slice(1);
      else if (content.startsWith('^')) content = '\\' + content;
      regex += '[' + content + ']'; i = end;
    } else regex += c.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  }
  try { return new RegExp(regex + '$', 's').test(value); } catch { return false; }
}
export class EventBus {
  private readonly listeners = new Map<string, Listener>();
  subscribe(listener: Listener): void {
    if (![listener.id, listener.business, listener.description].every(v => typeof v === 'string' && v.trim()) || !listener.routes?.length) throw new TypeError('listener requires metadata and routes');
    if (typeof listener.handler !== 'function' || !['active', 'noop'].includes(listener.mode || 'active')) throw new TypeError('invalid listener handler or mode');
    for (const values of [listener.routes, listener.exclude || [], listener.providers || [], listener.repositories || []]) {
      if (!Array.isArray(values) || values.some(v => typeof v !== 'string' || !v)) throw new TypeError('invalid listener selectors');
    }
    if (this.listeners.has(listener.id)) throw new TypeError('duplicate listener id');
    this.listeners.set(listener.id, Object.freeze({ ...listener, routes: Object.freeze([...listener.routes]), exclude: Object.freeze([...(listener.exclude || [])]),
      providers: Object.freeze([...(listener.providers || [])]), repositories: Object.freeze([...(listener.repositories || [])]), enabled: listener.enabled ?? true, mode: listener.mode || 'active' }));
  }
  inventory(): Doc[] {
    return [...this.listeners.values()].map(({ handler: _handler, ...l }) => structuredClone({ ...l, mode: l.enabled ? l.mode : 'disabled' }));
  }
  async dispatch(event: BotEvent): Promise<Outcome> {
    const route = routeOf(event);
    const matching = [...this.listeners.values()].filter(l => l.enabled && l.routes.some(p => glob(route, p)) && !l.exclude!.some(p => glob(route, p)) &&
      (!l.providers!.length || l.providers!.includes(event.provider)) && (!l.repositories!.length || l.repositories!.some(r => r.toLowerCase() === event.repo.toLowerCase())));
    const result = outcome(route); result.handled = matching.length > 0;
    for (const listener of matching) {
      try {
        // Each subscriber owns its copy; a mutating extension cannot corrupt later consumers or the archive.
        const reply = await listener.handler(structuredClone(event));
        if (reply !== undefined && reply !== null && (typeof reply !== 'object' || Array.isArray(reply))) throw new TypeError('listener must return an object or undefined');
        if (reply?.hook && reply.method) result.hooks.push(`${reply.hook}.${reply.method}`);
        result.listeners.push({ id: listener.id, status: String(object(reply).status || 'ok') });
      } catch (error) {
        const kind = category(error);
        result.errors.push(`${listener.id}: ${kind}`); result.listeners.push({ id: listener.id, status: 'error', error: kind });
      }
    }
    return result;
  }
}
export class NoopBoard implements Board {
  readonly mode = 'noop' as const; readonly repositories: readonly string[] = [];
  async handle(method: string, _event: BotEvent): Promise<Doc> { return { hook: 'board', method, status: 'noop' }; }
  status(): Doc { return { enabled: false }; }
}
export function registerBuiltin(bus: EventBus, board: Board): void {
  const analyze: [string, string[], string][] = [
    ['on_issue', ['issue.opened', 'issue.edited', 'issue.reopened'], 'Issue 新建、编辑和重新打开的分析入口；当前仅记录调用。'],
    ['on_issue_comment', ['issue_comment', 'issue_comment.*'], 'Issue / PR 评论分析入口；当前仅记录调用。'],
    ['on_pull_request', ['opened', 'synchronize', 'reopened', 'edited', 'ready_for_review'].map(a => `pull_request.${a}`), 'PR 变更分析入口；当前仅记录调用。'],
    ['on_pull_request_review', ['pull_request_review', 'pull_request_review.*'], 'PR 评审和行内评论入口；当前仅记录调用。'],
    ['on_pull_request_merged', ['pull_request.merged'], 'PR 合并后的分析入口；当前仅记录调用。'],
  ];
  for (const [method, routes, description] of analyze) bus.subscribe({ id: `analyze.${method}`, business: '内容分析', description, routes, mode: 'noop', handler: () => ({ hook: 'analyze', method, status: 'noop' }) });
  const boards: [string, string[], string[]?][] = [
    ['on_issue', ['issue', 'issue.*']], ['on_issue_comment', ['issue_comment', 'issue_comment.*']],
    ['on_pull_request', ['pull_request', 'pull_request.*', 'pull_request_review', 'pull_request_review.*'], ['pull_request.merged']],
    ['on_pull_request_merged', ['pull_request.merged']], ['on_push', ['push', 'push.*']],
    ['on_quality', QUALITY_KINDS.flatMap(kind => [kind, `${kind}.*`])],
  ];
  for (const [method, routes, exclude] of boards) bus.subscribe({ id: `board.${method}`, business: '看板更新', description: board.mode === 'active' ? '按源仓排队更新对应静态看板。' : '未启用静态发布，当前仅记录调用。',
    routes, exclude, mode: board.mode, providers: board.mode === 'active' ? ['github'] : [], repositories: board.repositories, handler: event => board.handle(method, event) });
}
export class Router {
  readonly bus = new EventBus();
  constructor(readonly board: Board = new NoopBoard()) { registerBuiltin(this.bus, board); }
  async dispatch(input: BotEvent): Promise<Outcome> {
    const event = input.kind === 'pull_request' && input.merged ? { ...input, action: 'merged' } : input;
    if (['ping', 'installation'].includes(event.kind)) return { ...outcome(routeOf(event), event.kind === 'ping' ? 'pong' : 'installation change recorded; no business listeners'), handled: true };
    const result = await this.bus.dispatch(event);
    if (!result.handled) result.note = 'no matching listener, recorded only';
    return result;
  }
}

File: sdbot/bus.py
"""In-process event bus; registered subscriptions are also the admin inventory."""
from __future__ import annotations

from dataclasses import dataclass, field
from fnmatch import fnmatchcase
import logging
import threading
from typing import Callable

from .events import Event


@dataclass
class Outcome:
    handled: bool
    route: str
    hooks: list[dict] = field(default_factory=list)
    errors: list[str] = field(default_factory=list)
    duplicate: bool = False
    note: str = ""
    listeners: list[dict] = field(default_factory=list)

    def hook_names(self) -> list[str]:
        return [f"{h.get('hook')}.{h.get('method')}" for h in self.hooks]


@dataclass(frozen=True)
class Listener:
    id: str
    business: str
    description: str
    routes: tuple[str, ...]
    handler: Callable[[Event], dict | None] = field(repr=False, compare=False)
    exclude: tuple[str, ...] = ()
    providers: tuple[str, ...] = ()
    repositories: tuple[str, ...] = ()
    enabled: bool = True
    mode: str = "active"

    def matches(self, event: Event) -> bool:
        return (self.enabled
                and any(fnmatchcase(event.route, pattern) for pattern in self.routes)
                and not any(fnmatchcase(event.route, pattern) for pattern in self.exclude)
                and (not self.providers or event.provider in self.providers)
                and (not self.repositories or event.repo.lower() in {r.lower() for r in self.repositories}))

    def public(self) -> dict:
        return {"id": self.id, "business": self.business, "description": self.description,
                "routes": list(self.routes), "exclude": list(self.exclude), "providers": list(self.providers),
                "repositories": list(self.repositories), "enabled": self.enabled,
                "mode": self.mode if self.enabled else "disabled"}


class EventBus:
    def __init__(self, log=None):
        self.log = log or logging.getLogger("sdbot.bus")
        self._listeners: dict[str, Listener] = {}
        self._lock = threading.RLock()

    def subscribe(self, listener: Listener) -> None:
        if not all(isinstance(v, str) and v.strip() for v in (listener.id, listener.business, listener.description)) or not listener.routes:
            raise ValueError("listener requires an id, business, description and routes")
        if not callable(listener.handler) or listener.mode not in ("active", "noop"):
            raise ValueError("invalid listener handler or mode")
        for selectors in (listener.routes, listener.exclude, listener.providers, listener.repositories):
            if not isinstance(selectors, tuple) or any(not isinstance(v, str) or not v for v in selectors):
                raise ValueError("listener selectors must be tuples of non-empty strings")
        with self._lock:
            if listener.id in self._listeners:
                raise ValueError("duplicate listener id")
            self._listeners[listener.id] = listener

    def inventory(self) -> list[dict]:
        with self._lock:
            return [listener.public() for listener in self._listeners.values()]

    def dispatch(self, event: Event) -> Outcome:
        with self._lock:
            listeners = [listener for listener in self._listeners.values() if listener.matches(event)]
        outcome = Outcome(bool(listeners), event.route)
        for listener in listeners:
            try:
                result = listener.handler(event)
                if result is not None and not isinstance(result, dict):
                    raise TypeError("listener must return a dict or None")
                if result and result.get("hook") and result.get("method"):
                    outcome.hooks.append(result)
                outcome.listeners.append({"id": listener.id, "status": str((result or {}).get("status", "ok"))})
            except Exception as err:
                # Exceptions may contain credentials or payloads. Retain only the category.
                category = type(err).__name__
                outcome.errors.append(f"{listener.id}: {category}")
                outcome.listeners.append({"id": listener.id, "status": "error", "error": category})
                self.log.warning("listener %s failed (%s)", listener.id, category)
        return outcome
